### ISSUE TITLE
Array based move generator

### ISSUE THREAD

**Issue by maksimKorzh** [maksimKorzh/xiangqi-stockfish#1]
@ianfab
Hi Fabien

First of all thank you so much for all the previous replies.

I have an array based movegen for xiangqi in js that I'm now porting to c++ and trying to integrate with sf.

I've realized that it's easier to write this custom movegen from scratch rather than having a pain of disabling all the bitboard related stuff.

My question is - is it possible to reuse search only (preferably without threads) from sf as a module?

I just don't need anything but search, uci and timing from sf.

Could you please kindly suggest the way to go?

Just a brief pointing of available directions/options would be more than enough. I just don't yet understand enough how much sf's search relies on a particular position object. It feels very monolithic and I want to make it as modular as possible.

Edit: what I currently do is writing function duplicates, for instance xq_set(fenStr) to init xiangqi board array from fen without touching bitboards. I can also make my own versions of generate(), pos.do_move() etc. And eventually end up on perft testing. But when it comes to search I feel very confused because apart from zobrist hash keys search seems to be relying on some additional keys that I neither want to maintain nor to inctementally update.
Hence I'm not sure whether my current approach is fine or leads to nowhere.

**Comment by ianfab** [comment]
I would see two possible ways to go:
1. If you are ready to burn bridges, removing everything related to position handling, move generation, and evaluation, and then rewriting a position class with the same interface (search uses plenty of methods of `Position`) as currently should work and might be the fastest theoretically.
2. I personally perhaps would not dare the first approach, because one can easily get lost. I guess I would first introduce the new data structures redundantly to the existing ones, and then step by step replace all usages of bitboards until none are left, and then remove them entirely. This has the advantage of always having a fully functional engine, and being able to easily check if a change breaks anything.

**Comment by maksimKorzh** [comment]
@ianfab
Thank you, first option seems more to fit my style)

Then probably the most essential question is how to debug?
Is there a way to run search without threads?
Is there a way to run separate functions from main.cpp?
(Say perft - it's available in sf7 but I have no idea how to call it from main.cpp in sf 12, so I called it via "go perft" UCI command)
Also the wat Position class is getting initialized is a tricky thing itself, for instance if seems like parse fen routine is called several times on init.

I do understand overall structure basing on my previous chess programming experience, but the feeling of
"something is going there that I can't control and keep track of" makes debugging very hard, say I'm getting
core dump error and it's not always that clear why that happens. I know debugger exists for this purpose but
even with a debugger without a clear understanding of how the class gets initialized it's hard to blindly modify it...

Fabien, can you briefly tell how did you start tinkering with sf's code?
Is there any documentation apart comment within the code?
SF feels very monolithic and not modular, do I feel it correctly?

Thanks in advance!

I promise not to torture you with stupid questions unless something technical and doesn't take long to answer,
I respect your time you're spending to answer me.

**Comment by ianfab** [comment]
Yes, you can put any function call in the main.cpp as long as you have the include statements to import them, if you want to do targeted debugging. However, you of course still might need to add required initialization, depending on what you want to test.

The initialization via FEN indeed is a bit tricky, since first the position of the UCI loop gets updated, and then an FEN is generated again and fed to the position objects of the threads. Therefore, if set() and fen() are not consistent, the search might get a different FEN than you put in.

The feeling of that it is somewhat hard to debug might be more related to C++ itself than the project specifically, at least in my experience. Especially memory and threading related issues can be hard to debug. That is why people created Rust ;). But I think asking such questions in the official SF forums as you now did is a good idea, since my specific knowledge with respect to chess variants is not really relevant for these general Stockfish/C++ topics, so other devs can perhaps help at least as good as me, at least as long as it is not Xiangqi specific.

PS: As a side note, my name is Fabi**a**n, but in English it sounds the same either way. And the German "ch" does not exist in many languages, but my last name can be quite well transcribed to cyrillyc as фіхтер if that helps you to pronounce it :).

**Comment by maksimKorzh** [comment]
@ianfab
Thank you.

re: PS
 - you probably watched my videos) Sorry, I'll try to pronounce your name properly from now on)

**Comment by maksimKorzh** [comment]
@ianfab 
Fabian, just wanted to thank you for your tips.
First I felt like I give up (countless number of times),
but then I've started following the approach you've suggested - disabling bitboards step by step.
At the moment I've changed the board array size.
I'm gonna be using Piece board[SQUARE_NB] for move generation,
now it's 11 * 14 squares - the board representation I've used in my Chinese chess engine wukong.
I've altered Position::set routine and the one printing the board to console and even if I run movegen or
perft I don't even get segfault!

However still getting segfault when trying to search but that should be because at very least mvv-lva sgould be broken after changining board size, number of files, ranks and adding advisor and cannon pieces.

The most fantastic thing is that I just feel like it flows well now and I do write code, debug further and further on.
I just never had experience working with someone else's code and this is really fun!

Also I already feel comfortable with a code base.
It feels like the same design patterns are applied to different modules which
helps to predict how it "might work" if you now how some other part works already.

I'm getting more and more enthusiasm with this project now.
I don't know if I eventually succeed but at least it's an exciting journey and
this has become possible thank to you.

Without your help I wouldn't be where I am now.
Thank you so much!

P.S. When I now reference Fairy Stockfish code I'm starting to realize how great the generalizations you've made are.
This is somewhat another level compared to what I am doing!
Just thank you for being available! 

A brief demo of current development state (parse fen/print board with enlarged board):
https://www.youtube.com/watch?v=inP7Ud3-G48&t=62s
